test/integration/apiserver/flowcontrol/fight_test.go

```diff
@@ -0,0 +1,219 @@
+	// maps FS key -> last written ResourceVersion
+	var lastRVs map[string]string = nil
+	// maps invert -> i -> FS key -> last notified ResourceVersion
+	notifiedRVs := map[bool][]map[string]string{
+		false: make([]map[string]string, size),
+		true:  make([]map[string]string, size)}
+	foreach := func(visit func(invert bool, i int)) {
+		for i := 0; i < size; i++ {
+			// The order of the following iteration is not deterministic,
+			// and that is good.
+			for invert := range ctlrs {
+				visit(invert, i)
+			}
+		}
+	}
+	foreach(func(invert bool, i int) {
+		myConfig := rest.CopyConfig(loopbackConfig)
+		myConfig = rest.AddUserAgent(myConfig, fmt.Sprintf("invert=%v, i=%d", invert, i))
+		myClientset := clientset.NewForConfigOrDie(myConfig)
+		fcIfc := myClientset.FlowcontrolV1beta1()
+		fsIfc := fcIfc.FlowSchemas()
+		if lastRVs == nil {
+			lastRVs = make(map[string]string)
+			// Wait until every FlowSchema is defined, and record its RV
+			allFlowSchemas := append(fcboot.MandatoryFlowSchemas, fcboot.SuggestedFlowSchemas...)
+			err := wait.Poll(time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				for _, fs := range allFlowSchemas {
+					_, err := fsIfc.Get(ctx, fs.Name, metav1.GetOptions{})
+					if err != nil {
+						return false, err
+					}
+					t.Logf("Got initial FlowSchema %#+v", fs)
+					key, _ := cache.MetaNamespaceKeyFunc(fs)
+					lastRVs[key] = fs.ResourceVersion
+				}
+				return true, nil
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+		informerFactory := informers.NewSharedInformerFactory(myClientset, 0)
+		fieldMgr := utilfc.ConfigConsumerAsFieldManager
+		foundToDangling := func(found bool) bool { return !found }
+		if invert {
+			fieldMgr = fieldMgr + "x"
+			foundToDangling = func(found bool) bool { return found }
+		}
+		notifiedRVs[invert][i] = map[string]string{}
+		ctlr := utilfc.NewTestable(utilfc.TestableConfig{
+			Name:  fmt.Sprintf("Controller%d[invert=%v]", i, invert),
+			Clock: clk,
+			FinishHandlingNotification: func(wq workqueue.RateLimitingInterface, obj interface{}) {
+				t.Logf("For invert=%v, i=%v, notified of %#+v", invert, i, obj)
+				obj = peel(obj)
+				switch typed := obj.(type) {
+				case *flowcontrol.FlowSchema:
+					key, _ := cache.MetaNamespaceKeyFunc(obj)
+					notifiedRVs[invert][i][key] = typed.ResourceVersion
+				}
+			},
+			AsFieldManager:         fieldMgr,
+			FoundToDangling:        foundToDangling,
+			InformerFactory:        informerFactory,
+			FlowcontrolClient:      fcIfc,
+			ServerConcurrencyLimit: 200,         // server concurrency limit
+			RequestWaitLimit:       time.Minute, // request wait limit
+			ObsPairGenerator:       metrics.PriorityLevelConcurrencyObserverPairGenerator,
+			QueueSetFactory:        fqtesting.NewNoRestraintFactory(),
+		})
+		ctlrs[invert][i] = ctlr
+		informerFactory.Start(stopCh)
+		if ctlr.WaitForCacheSync(stopCh) {
+			t.Logf("Achieved initial sync for invert=%v, i=%v", invert, i)
```

Comment:
You expressed surprise that this can print "late".  Since waiting is done for a short period and we assume it doesn't really matter how precise the handling is, it's a simple polling implementation, so it's not surprising that sometimes this print statement loses the race and you get a print from the controller first.

Reply:
The surprise is that there are _no_ print statements from the test notification handlers after the controllers start making changes.

Reply:
Argh, this should have been no surprise; the notification handler was reacting only if the spec changed.